### PR TITLE
Leaderboards

### PR DESCRIPTION
**Pull request by Ivanowicz**
Adds both a full course leaderboard for the lecturer and shows the student only his position on the leaderboard. No histogram yet.

### CODE REVIEW

File: src/main/java/upeldev/com/github/upel3/model/Course.java [hunk 1]
```diff
@@ -42,7 +42,7 @@
             joinColumns = @JoinColumn(name="COURSE_ID"),
             inverseJoinColumns = @JoinColumn(name="USER_ID")
     )
-    private Set<User> enrolledStudents = new HashSet<>();
+    public Set<User> enrolledStudents = new HashSet<>();
```
Comment (anosek-an) [inline src/main/java/upeldev/com/github/upel3/model/Course.java]: Why public? Lombok generates getters and setters automatically

**Review by anosek-an — COMMENTED** [review]
Other than the unnecessary modifier change it's fine. Handling different types of aggregation of points would be also nice, but I think it's best if we handle it in a separate issue.

**Comment by marcinz99** [comment]
Screenshot_2021-06-02 UPeL3 0
1. "leader" jako tytuł?
2. Sugerowałbym wyświetlanie imienia i nazwiska zamiast maila.
3. "Numer" -> "Pozycja"

File: src/main/java/upeldev/com/github/upel3/controllers/LeaderboardController.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        model.addAttribute("user", currentUser);
+
+        Set<User> users = currentCourse.getEnrolledStudents();
+
+        for (User user : users) {
+            List<Grade> grades = gradeService.findGradeByCourseAndUser(currentCourse, user);
+            sum = 0;
+            for (Grade grade : grades){
+                sum = sum + grade.getValue();
+            }
```
Comment (Krlstch) [inline src/main/java/upeldev/com/github/upel3/controllers/LeaderboardController.java]: This awkward loop could be replaced with:
double sum = grades.stream().mapToDouble(Grade::getValue).sum();

File: src/main/resources/templates/leaderboard_lecturer.html [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  <table class="table table-hover table-bordered bg-light">
+    <thead>
+    <tr>
+      <th scope="col">Numer</th>
+      <th scope="col">Student</th>
+      <th scope="col">Suma punktów</th>
+    </tr>
+    </thead>
+    <tbody>
+    <tr th:each="student : ${studentGrades}" th:class="${studentStat.count}">
```
Comment (marcinz99) [inline src/main/resources/templates/leaderboard_lecturer.html]: I'd add `th:style="${student.key} >= ${course.getPassValue()} ? 'background-color: #d0faa3;'"` to highlight green the ones that have already the course passed

File: src/main/resources/templates/leaderboard_lecturer.html [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  <link th:rel="stylesheet" th:href="@{/webjars/bootstrap/4.0.0-2/css/bootstrap.min.css} "/>
+  <link rel="shortcut icon" href="https://upel2.cel.agh.edu.pl/wiet/theme/image.php/celupel/theme/1619178409/favicon" />
+</head>
+<body style="background-color: #ecf1fc;">
+
+<div th:replace="fragments/header :: header"></div>
+
+<section class="container py-3">
+  <header class="d-flex flex-wrap align-items-center justify-content-center justify-content-md-between pt-4 pb-2 mb-0" style="border-top: 3px solid #343a40;">
+    <h2 class="my-0">leader</h2>
```
Comment (marcinz99) [inline src/main/resources/templates/leaderboard_lecturer.html]: leader -> 'Leaderboard'
Or even change it here and in course view to 'Tablica wyników' for linguistic consistency

**Review by anosek-an — APPROVED** [review]

**Pull request merged**